src/persist-client/src/cfg.rs

```diff
@@ -0,0 +1,278 @@
+            state_versions_recent_live_diffs_limit: usize::cast_from(
+                30 * Self::NEED_ROLLUP_THRESHOLD,
+            ),
+            writer_lease_duration: 60 * Duration::from_secs(60),
+            reader_lease_duration: Self::DEFAULT_READ_LEASE_DURATION,
+            critical_downgrade_interval: Duration::from_secs(30),
+            // TODO: This doesn't work with the process orchestrator. Instead,
+            // separate --log-prefix into --service-name and --enable-log-prefix
+            // options, where the first is always provided and the second is
+            // conditionally enabled by the process orchestrator.
+            hostname: std::env::var("HOSTNAME").unwrap_or_else(|_| "unknown".to_owned()),
+        }
+    }
+
+    /// Returns a new instance of [PersistConfig] for tests.
+    #[cfg(test)]
+    pub fn new_for_tests() -> Self {
+        use mz_build_info::DUMMY_BUILD_INFO;
+        use mz_ore::now::SYSTEM_TIME;
+
+        let mut cfg = Self::new(&DUMMY_BUILD_INFO, SYSTEM_TIME.clone());
+        cfg.hostname = "tests".into();
+        cfg
+    }
+}
+
+pub(crate) const MB: usize = 1024 * 1024;
+
+impl PersistConfig {
+    /// Default value for [`PersistConfig::blob_target_size`].
+    pub const DEFAULT_BLOB_TARGET_SIZE: usize = 128 * MB;
+    /// Default value for [`PersistConfig::compaction_minimum_timeout`].
+    pub const DEFAULT_COMPACTION_MINIMUM_TIMEOUT: Duration = Duration::from_secs(90);
+
+    // Move this to a PersistConfig field when we actually have read leases.
+    //
+    // MIGRATION: Remove this once we remove the ReaderState <->
+    // ProtoReaderState migration.
+    pub(crate) const DEFAULT_READ_LEASE_DURATION: Duration = Duration::from_secs(60 * 15);
+
+    // Tuning notes: Picked arbitrarily.
+    pub(crate) const NEED_ROLLUP_THRESHOLD: u64 = 128;
+}
+
+impl ConsensusKnobs for PersistConfig {
+    fn connection_pool_max_size(&self) -> usize {
+        self.consensus_connection_pool_max_size
+    }
+
+    fn connection_pool_ttl(&self) -> Duration {
+        self.consensus_connection_pool_ttl
+    }
+
+    fn connection_pool_ttl_stagger(&self) -> Duration {
+        self.consensus_connection_pool_ttl_stagger
+    }
+}
+
+/// Persist configuration parameters.
+///
+/// Parameters can be set (`Some`) or unset (`None`).
+/// Unset parameters should be interpreted to mean "use the previous value".
+#[derive(Clone, Debug, Default, PartialEq, Eq, Serialize, Deserialize, Arbitrary)]
+pub struct PersistParameters {
+    /// Configures [`PersistConfig::blob_target_size`].
+    pub blob_target_size: Option<usize>,
+    /// Configures [`PersistConfig::compaction_minimum_timeout`].
+    pub compaction_minimum_timeout: Option<Duration>,
+}
+
+impl PersistParameters {
+    /// Update the parameter values with the set ones from `other`.
+    pub fn update(&mut self, other: PersistParameters) {
+        if let Some(v) = other.blob_target_size {
```

Comment:
nit: I think that
```rust
if other.blob_target_size.is_some() {
    self.blob_target_size = other.blob_target_size;
}
```
spares one allocation without being substantially longer.

Reply:
this PR is a pure code movement PR, so I'm going to hold off on making any changes to this here. I am quite curious though, does that really save an allocation?

Reply:
I'm not 100% sure, just intuition. [Here is a minimal example](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=301e3034783523537954451aae3d9e86), probably the MIR and LLVM code of that is sufficient evidence if one is better than the other.

In any case I'm not sure over-optimizing this is worth it at the moment, it was just a suggestion.

Reply:
They seem to compile to the exact same assembly:
```asm
playground::Options::update1: # @playground::Options::update1
# %bb.0:
	cmpq	$1, %rsi
	jne	.LBB0_2
# %bb.1:
	movq	$1, (%rdi)
	movq	%rdx, 8(%rdi)

.LBB0_2:
	retq
                                        # -- End function
.set playground::Options::update2, playground::Options::update1
```

Reply:
Allocations are always explicit in Rust; i.e., you can't allocate without calling a function that ultimately calls a method on `Allocator` (or some equivalent unsafe C function like `malloc`).

Since there are no functions being called in this code, there can't be any allocations.

Reply:
I think the main confusion came from the fact that I didn't realize that both
```rust
self.blob_target_size = Some(v);
```
and
```rust
self.blob_target_size = other.blob_target_size;
```
just write the variant tag and value to the already allocated memory with the same number of instructions
```
# %bb.1:
	movq	$1, (%rdi)
	movq	%rdx, 8(%rdi)
```
For some reason I though that the latter is more efficient.